server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -6139,4 +6140,78 @@ public void testMaxDocsOnReplica() throws Exception {
+        final AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.NO_OPS_PERFORMED);
+        try (Store store = createStore()) {
+            EngineConfig config =
+                config(defaultSettings, store, createTempDir(), newMergePolicy(), null, null, globalCheckpoint::get);
+
+            store.createEmpty();
+            final String translogUUID = Translog.createEmptyTranslog(
+                config.getTranslogConfig().getTranslogPath(),
+                SequenceNumbers.NO_OPS_PERFORMED,
+                shardId,
+                primaryTerm.get()
+            );
+            store.associateIndexWithNewTranslog(translogUUID);
+
+            try (InternalEngine engine = createEngine(config)) {
+                engine.flush();
+                Map<String, String> userData = engine.getLastCommittedSegmentInfos().getUserData();
+                assertThat(userData, hasKey(ES_VERSION));
+                assertThat(userData.get(ES_VERSION), is(equalTo(Version.CURRENT.toString())));
+            }
+        }
+    }
+
+    public void testTrimUnsafeCommitHasESVersionInUserData() throws IOException {
+        final AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.NO_OPS_PERFORMED);
+        final int maxSeqNo = 40;
+        final List<Long> seqNos = LongStream.rangeClosed(0, maxSeqNo).boxed().collect(Collectors.toList());
+        Collections.shuffle(seqNos, random());
+        try (Store store = createStore()) {
+            EngineConfig config =
+                config(defaultSettings, store, createTempDir(), newMergePolicy(), null, null, globalCheckpoint::get);
+
+            try (InternalEngine engine = createEngine(config)) {
+                for (int i = 0; i < seqNos.size(); i++) {
+                    ParsedDocument doc = testParsedDocument(Long.toString(seqNos.get(i)),
+                        null,
+                        testDocument(),
+                        new BytesArray("{}"),
+                        null
+                    );
+                    Engine.Index index = new Engine.Index(newUid(doc),
+                        doc,
+                        seqNos.get(i),
+                        0,
+                        1,
+                        null,
+                        REPLICA,
+                        System.nanoTime(),
+                        -1,
+                        false,
+                        UNASSIGNED_SEQ_NO,
+                        0
+                    );
+                    engine.index(index);
+                    if (randomBoolean()) {
+                        engine.flush();
+                        final Long maxSeqNoInCommit = seqNos.subList(0, i + 1).stream().max(Long::compareTo).orElse(-1L);
+                    }
+                }
+                globalCheckpoint.set(randomInt(maxSeqNo));
+                engine.syncTranslog();
+            }
+
+            store.trimUnsafeCommits(config.getTranslogConfig().getTranslogPath());
```

Comment:
This feels odd, since `trim` used to copy the existing user data over anyway and es version has already been set. I wonder if we can reset the version randomly to a prior one using the same lucene version? It would at least validate this a bit, though it is also somewhat flaky, since there would be no validation after every lucene upgrade... Better ideas are welcome!